src/Distributor.sol

```diff
@@ -0,0 +1,126 @@
+
+    /// @notice Track claimed actions.
+    mapping(bytes32 => bool) public claimed;
+
+    /// @notice The Elixir signer address.
+    address public immutable signer;
+
+    /*//////////////////////////////////////////////////////////////
+                                 EVENTS
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Emitted when a user claims a token amount.
+    /// @param user The user who claimed the rewards.
+    /// @param token The token claimed.
+    /// @param amount The amount of rewards claimed.
+    /// @param nonce The nonce of the action.
+    event Claimed(address indexed user, address indexed token, uint256 indexed amount, uint256 nonce);
+
+    /// @notice Emitted when the owner withdraws a token.
+    event Withdraw(address indexed token, uint256 indexed amount);
+
+    /*//////////////////////////////////////////////////////////////
+                                 ERRORS
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Error emitted when the ECDSA signer is not correct.
+    error InvalidSignature();
+
+    /// @notice Error emitted when the user has already claimed.
+    error AlreadyClaimed();
+
+    /// @notice Error emitted when the amount is zero.
+    error InvalidAmount();
+
+    /// @notice Error emitted when the nonce is zero.
+    error InvalidNonce();
+
+    /// @notice Error emitted when the token is zero.
+    error InvalidToken();
+
+    /*//////////////////////////////////////////////////////////////
+                               CONSTRUCTOR
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Set the storage variables.
+    /// @param _name The name of the contract.
+    /// @param _version The version of the contract.
+    /// @param _signer The Elixir signer address.
+    constructor(string memory _name, string memory _version, address _signer) EIP712(_name, _version) {
+        // Set the signer address.
+        signer = _signer;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                           EXTERNAL FUNCTIONS
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Claims tokens approved by Elixir.
+    /// @param token The token to claim.
+    /// @param amount The amount of token to claim.
+    /// @param nonce The nonce of the action.
+    /// @param signature The signature from the Elixir signer.
+    function claim(address token, uint256 amount, uint256 nonce, bytes memory signature) external {
+        // Check that the token is not zero.
+        if (token == address(0)) revert InvalidToken();
+
+        // Check that the amount is not zero.
+        if (amount == 0) revert InvalidAmount();
+
+        // Check that the nonce is not zero.
+        if (nonce == 0) revert InvalidNonce();
+
+        // Generate digest.
+        bytes32 digest = _hashTypedDataV4(
+            keccak256(
```

Comment:
Any reason not to use the direct output of `keccak256` for the claim lookup?

Reply:
`_hashTypedDataV4` is for EIP 712, which is a security layer that adds params, like the chain id, to prevent issues, like using the same signature across multiple chains. it's a general good security practice.